### ISSUE TITLE
[nodes] Change or stop forcing camelCasing for unique names

### ISSUE THREAD

**Issue by lorisleiva** [codama-idl/codama#369]
This is a proposal that needs further discussion before implementing.

---

When designing the Codama standard, we needed to agree on a string casing for named nodes in order to compare them by names and went with `camelCase` to match the casing of node attributes.

The slight issue with `camelCase` is that it contains slightly less information than other casing such as `Title Case` or `snake_case` mainly due to number positioning. For instance, with `snake_case` we can define something like `token_2022_3rd_edition` without any ambiguity on the numbers 2022 and 3. On the other hand, `camelCase` cannot keep up with this distinction as it results in `token20223rdEdition` which re-`snake_cased` would give you `token20223rd_edition`.

The bottom line is, casings with explicit separators like `snake_case` keep more information than casing like `camelCase` and thus, using the former instead of the latter as a naming convention might make more sense.

This is a breaking change on the standard itself and would need to be a major version bump.

**Comment by lorisleiva** [comment]
> Does 'Token 2022 3rd Edition' actually exist?

Not this was just an example illustrating the issue but probably not the best one haha. I've already come across that issue a couple of time so it's not super rare but I agree it is an edge case. I just think that, if you're going to transform strings into a common casing strategy, we might as well use one that loose as little information as possible.

Another totally viable strategy would be to stop transforming string and leaving them as-is. The issue is matching nodes by name would become more prone to error but, at the end of the day, if you treat the `name` as some identifier then you know it must match exactly.

**Comment by stegaBOB** [comment]
If for whatever reason a name needs to be a pubkey, this can cause issues too. We have a separate intermediate IDL spec at star atlas for our framework that we convert to program nodes, and some of our items in the spec are namespaced by program id. Camel case causes issues with that too. We are able to refactor that issue out on our end, but thought I'd comment anyway in case that might be important for anyone else. Snake case wouldn't really fix that unfortunately.

**Comment by mikemaccana** [comment]
It seems reasonable that although these issues do occasionally happen, trying to solve the issues of casing on a platform is out of scope for Codama.

**Comment by lorisleiva** [comment]
Moved to https://github.com/codama-idl/spec/issues/129
